pkg/daemon/update.go

```diff
@@ -122,6 +122,29 @@ func (dn *Daemon) performPostConfigChangeAction(postConfigChangeActions []string
 	return dn.triggerUpdateWithMachineConfig(state.currentConfig, state.desiredConfig, true)
 }
 
+func setRunningKargsWithCmdline(config *mcfgv1.MachineConfig, requestedKargs []string, cmdline []byte) error {
+	splits := strings.Split(strings.TrimSpace(string(cmdline)), " ")
+	config.Spec.KernelArguments = nil
+	for _, split := range splits {
+		for _, reqKarg := range requestedKargs {
+			if strings.ReplaceAll(reqKarg, "\"", "") == strings.ReplaceAll(split, "\"", "") {
```

Comment:
I don't understand this line, what's the `ReplaceAll` doing here?  

I would say we should try to align with `validateKernelArguments`

Reply:
In /proc/cmdline the arguments appear without quotes even if the quotes were provided to ostree as kernel arguments.  To align them, the quotes are removed here.

Reply:
> even if the quotes were provided to ostree as kernel arguments

What quotes given to ostree?  The ostree side just passes through these things.  Where exactly are the quotes coming from?

Reply:
If we take the kernelArguments in the file `/etc/ignition-machine-config-encapsulated.json` we get:
```
[
  "systemd.unified_cgroup_hierarchy=1",
  "cgroup_no_v1=\"all\"",
  "psi=1"
]
```
But after they are applied, in th cmdline file, it looks like (partial):
```
systemd.unified_cgroup_hierarchy=1 cgroup_no_v1=all psi=1
```
So the second one appears without quotes in the cmdline file.

Reply:
OK let's be clear then, we're talking about `cgroup_no_v1="all"` as the weird outlier here.  Looks like that comes from this code: https://github.com/openshift/machine-config-operator/blob/5b821a279c88fee1cc1886a6cf1ec774891a2258/pkg/controller/kubelet-config/helpers.go#L180

Which...what the heck is that?  (a bit of time passes) OK, I found it in https://docs.kernel.org/admin-guide/cgroup-v2.html

But I think whoever wrote that code was likely confused...it certainly seems to me like the Linux kernel is intending to parse `cgroup_no_v1=all`, not `cgroup_no_v1="all"`.  (Though I bet I see why the code author made that mistake because the kernel *docs* seem to include the quotes in one case but not another).

Compounding the fun here, also AFAICS there are zero warnings for unknown/unhandled parameters here.  So specifying `cgroup_no_v1="all"` just silently does nothing (edit: Nope, see below).

Now the *next* large problem here is that *something* in the stack is "eating" these quotes.

Playing with this using e.g.:

```
$ rpm-ostree kargs --append='cgroup_no_v1="all"'
$ reboot
...
$ cat /proc/cmdline
... cgroup_no_v1=all
```

However, notice the bootloader entry *does* have those quotes:

```
$ grep options /boot/loader/entries/ostree-2-rhcos.conf 
options ignition.platform.id=qemu ... cgroup_no_v1="all" 
$
```

If I was a betting person here (and I am) then my money would be on grub doing the wrong thing.

Wait, wait no perhaps not: I think it's pretty clear, check out this kernel code:
https://github.com/torvalds/linux/blob/7ba2090ca64ea1aa435744884124387db1fac70f/lib/cmdline.c#L224

It's clearly implementing a minimal subset of "shell like" parsing...

> but can't escape "

Oh man, what an ad-hoc ill-specified mess!

Reply:
Offhand I think this logic dates back to the Linux kernel pre-history https://github.com/torvalds/linux/commit/1da177e4c3f41524e886b7f1b8a0c1fc7321cac2

So here's what I'd say:

- Factor out a helper function named e.g. `parseKernelArgumentsLikeLinux` that performs the same steps as the Linux kernel in "parsing" command line arguments  (and I'm now realizing that we probably need to do this in ostree side too)
- Cross-reference the Linux kernel code
- Consider changing the MCO code to drop the unnecessary quotes because we've just painted ourselves into an unfortunate corner case with this


*Alternatively* - notice that `rpm-ostree kargs` outputs what it wrote into the bootloader config!  So instead of parsing `/proc/cmdline`, we could parse that (which again is what the other MCO code is doing)

Reply:
Using `rpm-ostree kargs` sounds like a good solution.